src/runtime/hexagon_remote/elf.h

```diff
@@ -0,0 +1,914 @@
+                // manual).
+                if (iclass < 3 || iclass > 7) {
+                    fail(__LINE__);
+                    return;
+                }
+
+                // Pull out the subinstructions. They're the low 13
+                // bits of each half-word.
+                uint32_t hi = (inst >> 16) & ((1 << 13) - 1);
+                uint32_t lo = inst & ((1 << 13) - 1);
+
+                // We only understand the ones where hi starts with 010
+                if ((hi >> 10) != 2) {
+                    fail(__LINE__);
+                    return;
+                }
+
+                // Low 6 bits of val go in the following bits.
+                mask = 63 << 20;
+
+            } else if ((inst >> 24) == 72) {
+                // Example instruction encoding that has this high byte (ignoring bits 1 and 2):
+                // 0100 1ii0  000i iiii  PPit tttt  iiii iiii
+                if (debug) log_printf("Instruction-specific case A\n");
+                mask = 0x061f20ff;
+            } else if ((inst >> 24) == 73) {
+                // 0100 1ii1  000i iiii  PPii iiii  iiid dddd
+                if (debug) log_printf("Instruction-specific case B\n");
+                mask = 0x061f3fe0;
+            } else if ((inst >> 24) == 120) {
+                // 0111 1000  ii-i iiii  PPii iiii  iiid dddd
+                if (debug) log_printf("Instruction-specific case C\n");
+                mask = 0x00df3fe0;
+            } else if ((inst >> 16) == 27209) {
+                // 0110 1010  0100 1001  PP-i iiii  i--d dddd
+                mask = 0x00001f80;
+            } else if ((inst >> 25) == 72) {
+                // 1001 0ii0  101s ssss  PPii iiii  iiid dddd
+                // 1001 0ii1  000s ssss  PPii iiii  iiid dddd
+                mask = 0x06003fe0;
+
+            } else {
+                log_printf("Unhandled!\n");
+                fail(__LINE__);
+                return;
+            }
+        }
+
+        uintptr_t old_val = val;
+        bool consumed_every_bit = false;
+        for (int i = 0; i < 32; i++) {
+            if (mask & (1 << i)) {
+                if (inst & (1 << i)) {
+                    // This bit should be zero in the unrelocated instruction
+                    fail(__LINE__);
+                    return;
+                }
+                // Consume a bit of val
+                int next_bit = val & 1;
+                if (is_signed) {
+                    consumed_every_bit |= ((intptr_t)val) == -1;
+                    val = ((intptr_t)val) >> 1;
+                } else {
+                    val = ((uintptr_t)val) >> 1;
+                }
+                consumed_every_bit |= (val == 0);
+                inst |= (next_bit << i);
+            }
+        }
+
+        if (verify && !consumed_every_bit) {
+            fail(__LINE__);
+            log_printf("Relocation overflow inst=%08lx mask=%08lx val=%08lx\n", inst, mask, (unsigned long)old_val);
```

Comment:
To avoid confusion in the log, it might be helpful to swap the order of these two (printf before fail), to be consistent with the other log + fail reports.